src/Xamarin.Android.Build.Tasks/Tasks/GenerateTypeMaps.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+using Microsoft.Build.Framework;
+
+using Java.Interop.Tools.Cecil;
+using Java.Interop.Tools.Diagnostics;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+    public class GenerateTypeMaps : AndroidTask
+	{
+		public override string TaskPrefix => "GTM";
+
+		[Required]
+		public ITaskItem[] ResolvedAssemblies { get; set; }
+
+		[Required]
+		public ITaskItem[] ResolvedUserAssemblies { get; set; }
+
+		[Required]
+		public ITaskItem [] FrameworkDirectories { get; set; }
+
+		[Required]
+		public string[] SupportedAbis { get; set; }
+
+		[Required]
+		public string OutputDirectory { get; set; }
+
+		[Required]
+		public bool GenerateNativeAssembly { get; set; }
+
+		[Output]
+		public string[] GeneratedBinaryTypeMaps { get; set; }
+
+		public bool ErrorOnCustomJavaObject { get; set; }
+
+		public override bool RunTask ()
+		{
+			try {
+				Run ();
+			} catch (XamarinAndroidException e) {
+				Log.LogCodedError (string.Format ("XA{0:0000}", e.Code), e.MessageWithoutCode);
+				if (MonoAndroidHelper.LogInternalExceptions)
+					Log.LogMessage (e.ToString ());
+			}
+
+			if (Log.HasLoggedErrors) {
+				// Ensure that on a rebuild, we don't *skip* the `_GenerateJavaStubs` target,
+				// by ensuring that the target outputs have been deleted.
+				Files.DeleteFile (Path.Combine (OutputDirectory, "typemap.index"), Log);
+				foreach (string file in Directory.EnumerateFiles (OutputDirectory, "*.typemap")) {
+					Files.DeleteFile (file, Log);
+				}
+			}
+
+			return !Log.HasLoggedErrors;
+		}
+
+		void Run ()
+		{
+			var interestingAssemblies = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
+
+			var res = new DirectoryAssemblyResolver (this.CreateTaskLogger (), loadDebugSymbols: true);
+			foreach (var dir in FrameworkDirectories) {
+				if (Directory.Exists (dir.ItemSpec))
+					res.SearchDirectories.Add (dir.ItemSpec);
+			}
```

Comment:
This adds an additional pass using Mono.Cecil to load every assembly.

Should we just put this code in the `<GenerateJavaStubs/>` task for now on this PR? It might be easier to leave to someone else (probably end up being me 🤦‍♂) to split that task up.

Otherwise, you might have to deal with some incremental build or performance tests failing. We'll see?

Reply:
Let's keep that as an option. I'll keep it separate for the moment, to see if everything works.

Reply:
When I last attempted to split up `<GenerateJavaStubs/>`, it hurt performance a bit, I had to move the `DirectoryAssemblyResolver` to be stored in `RegisterTaskObject` so it could be shared across multiple tasks:

https://github.com/xamarin/xamarin-android/compare/master...jonathanpeppers:split-generatejavastubs

This branch isn't quite finished, the `AndroidManifest.xml` portion is broken.

Reply:
It would be nice to know what the build performance impact will be. I see you have TODO items in the PR comment to include those in the commit messages. Lets see how much things chance and decide if we need to move the code then.

Reply:
See https://github.com/xamarin/xamarin-android/pull/3992#discussion_r366824141 :)